playbooks/roles/mongodb/tasks/main.yml

```diff
@@ -70,12 +70,47 @@
 - name: Install Prometheus MongoDB exporter
   when: MONGODB_EXPORTER_ENABLED
   block:
+    - name: Create custom fact directory
+      file:
+        path: "/etc/ansible/facts.d"
+        state: "directory"
 
-    - name: Download MongoDB exporter
-      get_url:
-        url: "{{ MONGODB_EXPORTER_DOWNLOAD_URL }}"
-        dest: /usr/local/bin/mongodb_exporter
-        mode: 0755
+    - name: Store MongoDB exporter version
+      copy:
+        content: |
+          {"version": "{{ MONGODB_EXPORTER_VERSION }}"}
+        dest: /etc/ansible/facts.d/mongodb_exporter.fact
+      register: mongodb_exporter_version
+
+    - name: Upgrade MongoDB exporter
+      when: mongodb_exporter_version.changed
```

Comment:
The idea with local facts is that they are loaded automatically by Ansible.  You can access the fact as `ansible_local.mongodb_exporter.version`.

The problem with the current approach is that it is quite fragile.  The new version is written first, before the actual installation is performed.  If there is an error with some task during the installation, fixing the error and re-running the playbook won't have any effect – the fact file already contains the new version, so nothing will happen.

Instead, you should simply query `ansible_local.mongodb_exporter.version` before the installation.  It will be either unset, or contain the currently installed version.  If it is unset, or if the installed version differs from the desired version, we can proceed with the installation.  After the installation, we can update the fact file, so it only gets updated when the installation is successful.

Reply:
Fair enough! However, trying to query `ansible_local.mongodb_exporter.version` before setting it up will cause a key error as Ansible not able to access the key from the fact. Will check how to overcome this and will implement the best practice.

Reply:
I think you should be able to use a condition like

    when: ansible_local.mongodb_exporter.version is not defined or ansible_local.mongodb_exporter.version != MONGODB_EXPORTER_VERSION